% Created 2012-11-30 Fri 14:43
\documentclass[11pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{fixltx2e}
\usepackage{graphicx}
\usepackage{longtable}
\usepackage{float}
\usepackage{wrapfig}
\usepackage{soul}
\usepackage{textcomp}
\usepackage{marvosym}
\usepackage{wasysym}
\usepackage{latexsym}
\usepackage{amssymb}
\usepackage{hyperref}
\tolerance=1000
\providecommand{\alert}[1]{\textbf{#1}}

\title{Scythe: A tool for removing 3'-end adapter contaminants using Bayesian classification}
\author{Vince Buffalo}
\date{}
\hypersetup{
  pdfkeywords={},
  pdfsubject={},
  pdfcreator={Emacs Org-mode version 7.8.11}}

\begin{document}

\maketitle

\setcounter{tocdepth}{3}
\tableofcontents
\vspace*{1cm}


\textbf{Motivation:} Modern sequencing technologies can leave artifactual
 contaminant sequences at the 3'-end of reads. 3'-end regions also
 have the lowest quality bases that are likely to be called
 incorrectly, which makes identifying and removing 3'-end contaminants
 difficult. Fixed-number mismatch approaches to remove contaminants
 perform poorly in these low quality regions. Failing to remove such
 contaminants can seriously confound downstream analyses like assembly
 and mapping.


\textbf{Results:} Scythe is a program designed specifically to remove 3'-end
 contaminants. It searches for 3'-end contaminants and uses a Bayesian
 model that considers individual base qualities to decide whether a
 given match is a contaminant or background sequence. Even for a
 variety of prior contamination rates, Scythe outperforms other
 adapter removal software tools.

\textbf{Availability:} Scythe is freely available under the MIT license at
 \href{https://github.com/vsbuffalo/scythe}{https://github.com/vsbuffalo/scythe}.

\section{Introduction}
\label{sec-1}


Scythe focuses on 3'-end contaminants, specifically those due to
adapters or barcodes. It embraces the Unix Philosophy of ``programs
that do one thing well''
(\href{http://www.faqs.org/docs/artu/ch01s06.html}{http://www.faqs.org/docs/artu/ch01s06.html}). Many second-generation
sequencing technologies such as Illumina's Genome Analyzer II and
HiSeq have lower-quality 3'-end bases. These low-quality bases are
more likely to have nucleotides called incorrectly, making contaminant
identification more difficult. Futhermore, 3'-end quality
deterioration is not uniform across all reads (see figure 1 in
Supplementary Materials), there is variation in the quality per base.

A common step in read quality improvement procedures is to remove
these low-quality 3'-end sequences from reads. This is thought to
increase mapping rates and improve assembly quality. However doing
quality-based 3'-end trimming before contaminant removal would remove
sequence that could be used (despite being unreliable) to identify the
contaminants more positively. Scythe takes the approach that it is
better to use full information, even if it's unreliable. How
unreliable a base is is indicated by the FASTQ quality score, which
can be incorporated into classification procedures.

Fixed-number of mismatch approaches have the disadvantage that they
don't differentially weight a mismatch on a low-quality base from a
mismatch on a high-quality base. Futhermore, the fixed-number could
easily be exhausted in a run of bad bases (which are quite common in
the 3'-end), even though every good-quality base perfectly matches the
contaminant sequence.
\section{Scythe's Methods}
\label{sec-2}


Scythe uses Bayesian methods to identify and remove a given set of
3'-end contaminants (most often sequencing adapters). Scythe only
checks for the adapters in the 3'-end; contaminants further towards
the middle and 5'-ends often have high quality bases, so identifying
and removing them is much simpler. Scythe makes some assumptions: see
the Supplementary Materials. TODOREF
\subsection{String matching in Scythe}
\label{sec-2-1}


For each adapter in the file, Scythe looks for the best match in
terms of scores. A nucleotide match is scored as a 1, and a mismatch
is scored as a -1. Because Scythe doesn't address contaminants with
insertions or deletions, it doesn't use a standard alignment strategy
(i.e. dynamic programming). Instead, it considers every possible
substring of contamination: for a contaminant of length $l_c$, it
scores all possible substrings from the 5'-end of the contaminant. The
top scoring match is then used in the classification procedure. For
efficiency, other matches are not used in the classification
procedure.\footnote{This option may be added to further Scythe versions.
 }

The time complexity of Scythe's match algorithm for a single adapter
of length is $l_a$ \$O(l\_{}a\^{}2 * R)\$ for a FASTQ file with $R$ entries.
\subsection{Bayesian classification of top-scoring matches}
\label{sec-2-2}


There are two mutually exclusive and exhaustive events: a top scoring
match is a contaminant or it is random sequence (that happens to be
similar to the adapter contaminant). A likelihood for each of these
events, given probabilities of each base being called correctly and
which bases mismatch the contaminant sequence, can be calculated.

These likelihood functions assume models of error for each event. If
the top-scoring match is contaminant sequence (event $C$), the
likelihood $P(S | C)$ (where $S$ is the sequence match data) is:

$$ P(S | C) = \prod_i^{l_t} q_i^{m_i} \cdot (1-q_i)^{1 - m_i} $$

Where $m_i \in \{0, 1\}$ indicating whether the $i$ base is a
mismatch or match (respectively) and $q_i$ is the probability the $i$
base is called correctly (from the ASCII-encoded quality). $l_t$ is
the length of the top-scoring match.

If the top-scoring sequence is not a contaminant sequence (event
$C'$), it is assumed that the matches are just by chance. Thus, the
likelihood function is:

$$ P(S | C') = \prod_i^{l_t} \left(\frac{1}{4}\right)^{m_i} \cdot \left(\frac{3}{4}\right)^{1 - m_i} $$

These likelihoods can then be combined using Bayes' Thereom to give
the probability of contamination given then top-scoring match:

$$ P(C|S) = \frac{P(C) P(S|C)}{P(S)} $$

Where the denominator can be replaced with $P(S) = P(S | C)P(C) +
P(S | C') P(C')$ by the Law of Total Probability. $P(C'|S)$ is
calculated similarly. Since these are mutually exclusive and
exhaustive events, the \emph{maximum a posteriori} rule can be used to
classify a top-scoring sequence as either contaminated or
non-contaminated (e.g. if $P(C'|S) > P(C|S)$, the sequence is not
contaminated and visa-versa).

Required in this Bayesian formulation is the prior of contamination,
$P(C)$. Specifying the prior may seem like a nuisance, but it allows
ultimately allows for more accurate classification in a variety of
different contamination scenarios. The prior doesn't need to be
estimated increadibly well; one technique that works well is to view
the FASTQ file in the Unix command line tool less and search for the
5'-end bases of the adapter contaminant. The number of results on a
page of less output divided by the number of FASTQ entries on that
page works well as an initial guess for the prior.
\section{Results}
\label{sec-3}


Scythe was tested against two similar program: Btrim (Kong, 2011) and
Cutadapt (Martin, 2011). Btrim

Cutadapt has an advantage over Scythe in that it does gapped
alignments (originally it was developed to trim 454 sequences which
have homopolymer repeats).


\bibliographystyle{plain}
\bibliography{references}

\end{document}